### ISSUE TITLE
🐛  Missing CHANGELOG Entries

### ISSUE THREAD

**Issue by unicornware** [flex-development/grease#26]
## Description

**All** commits for a new project version are not added to its `CHANGELOG` entry.

## Steps to reproduce

1. Make note of recent commits
2. Run `grease` with `options.dryRun === true` in a project versioned after `v1.0.0`
3. See error in `changelog` lifecycle `options.dryRun` output

## Expected behavior

With respect to `options.path`, all commits should be able to be found in the `CHANGELOG`.

## Relevant screenshots

**Commits that made it into the `CHANGELOG` interspersed with missing `build` scope commits**

![missing-changelog-entries-1](https://user-images.githubusercontent.com/20430108/133641100-cfb23fd0-26a2-4bde-9c4f-418b1434f592.png)

**`grease` output**

![missing-changelog-entries-2](https://user-images.githubusercontent.com/20430108/133641146-7af7d618-1135-4d76-a14e-46f2dd08297b.png)

## Environment

- Node Version: `v16.6.0`
- OS: `macOS v11.3.1`
- Package Manager: `yarn@3.0.2`
- Package version(s): `grease@1.1.0`

## Possible solution

Unsure, need to search the [`standard-version`][1] library for possible [related issues][2] and debug `grease`.

## Additional context

- `options.path` was unset
- Unsure if the issue is reproducible (may be related to #24)
- During this particular run, the `CHANGELOG` entry for `@flex-development/log@2.0.0` was missing all `build` (:hammer: Build) scope commits. [Commit entries][3] needed to be manually added

[1]: https://github.com/conventional-changelog/standard-version
[2]: https://github.com/conventional-changelog/standard-version/issues?q=
[3]: https://github.com/flex-development/log/blob/v2.0.0/CHANGELOG.md#hammer-build

**Comment by unicornware** [comment]
Reproduced error in the `grease` repo. 

Commits after [`3d97289`][1] under the path `$PWD/packages/grease` are not included in the `CHANGELOG` entry for `grease@1.1.1`:

```zsh
▲ (FLEXBOOK) FLDV-P010 [next] yarn release:grease --release-as=patch --dry-run 
! starting release workflow @flex-development/grease [dry=true]
✔ Running lifecycle script "prerelease"
ℹ - execute command: "yarn workspace @flex-development/grease test --no-cache"
✔ bumping version in package.json from 1.1.0 to 1.1.1
✔ outputting changes to CHANGELOG.md

---
### [1.1.1](https://github.com/flex-development/grease/compare/grease@1.1.0...grease@1.1.1) (2021-09-18)


### :pencil2: Housekeeping

* add eslint base config ([8de32f3](https://github.com/flex-development/grease/commit/8de32f3e4dd32c7f1cea85c9943b7ec88a1d7153))
* refine commit scopes ([ae15022](https://github.com/flex-development/grease/commit/ae15022381725668a773ab09c5c38b784e7cd83d))
* run `yarn set version berry` ([f94a4e1](https://github.com/flex-development/grease/commit/f94a4e162dbc3de7038b1bdb79573e25427baa5d))
* **typescript:** add typings for `shell-quote@1.7.2` ([70156cb](https://github.com/flex-development/grease/commit/70156cb7751374830f5c74d5b3c9d9f58c4b3e6a))
* update environment variables ([3f55036](https://github.com/flex-development/grease/commit/3f55036bf343aa8a3079294d3ebb4158cb4068e4))
* use yarn workspace based commit scopes ([889322b](https://github.com/flex-development/grease/commit/889322b2d5224a58dbde522f81cc3177da981c86))
---
```

After some additional debugging, however, I found that the [`git-raw-commits`][2]  library **_does_** find all commits as expected:

```zsh
▲ (FLEXBOOK) FLDV-P010 [next] git-raw-commits --from grease@1.1.0 --path $PWD/packages/grease
build(deps-dev): use `@types/faker@5.5.8`

build(deps): use `isomorphic-git@1.10.0`

build(deps): use `lodash.*` deps

instead of `lodash/*`

build(deps-dev): drop `@types/yargs-unparser@2.0.1`

build(deps-dev): move `@types/lodash@4.14.172` to `devDependencies`

build(deps-peer): add `@types/shelljs@0.8.9`

build(deps-dev): move `@types/shelljs@0.8.9` to `devDependencies`

build(deps-peer): add `@types/node>=15`

build(deps-dev): use `@types/node@16.9.2`

test(grease): update changelog fixtures

test(grease): use `DTO` fixture in `IsPath` functional test

refactor(grease): `CreateReleaseDTO#toString`

P010-33

chore(typescript): add typings for `shell-quote@1.7.2`

P010-32

build(deps): add `shell-quote@1.7.2`

P010-32

refactor(grease): integrate `@flex-development/log@2.0.0`

build(deps): add `@flex-development/log@2.0.0`

chore: refine commit scopes

build: run local integrity check

chore: use yarn workspace based commit scopes

chore: update environment variables

chore: add eslint base config

chore: run `yarn set version berry`

build: update build output
```

Still in the process of tracking down the root issue.

[1]: https://github.com/flex-development/grease/commit/3d972890cfa69e9f9622a54dc5371930d7079131
[2]: https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/git-raw-commits

**Comment by unicornware** [comment]
**The root issue is how `grease` merges default options: `args = merge({}, defaults, args)`**.

The default value for `IGreaseOptions.types` is:

```typescript
[
  { type: 'feat', section: 'Features' },
  { type: 'fix', section: 'Bug Fixes' },
  { type: 'chore', hidden: true },
  { type: 'docs', hidden: true },
  { type: 'style', hidden: true },
  { type: 'refactor', hidden: true },
  { type: 'perf', hidden: true },
  { type: 'test', hidden: true }
]
```

`greases` uses [`lodash.merge`][1] to merge default options. If a project maintainer sets `IGreaseOptions.types` to

```typescript
[
  { type: 'feat', section: ':sparkles: Features' },
  { type: 'fix', section: ':bug: Fixes' },
  { type: 'revert', section: ':rewind: Revert' },
  { type: 'test', section: ':robot: Testing' },
  { type: 'docs', section: ':book: Documentation' },
  { type: 'build', section: ':hammer: Build' },
  { type: 'refactor', section: ':recycle: Code Improvements' },
  { type: 'perf', section: ':zap: Performance Updates' },
  { type: 'style', section: ':nail_care: Formatting & Structure' },
  {
    type: 'ci',
    section: ':truck: Continuous Integration & Deployment'
  },
  { type: 'chore', section: ':pencil2: Housekeeping' },
  { type: 'wip', hidden: true }
]
```

commit types with a `section` field aren't visible in the `CHANGELOG` as expected. Due to the merge, some fields even have a `hidden` field appended. The resulting `types` config after merge is:

```typescript
[
  { type: 'feat', section: ':sparkles: Features' },
  { type: 'fix', section: ':bug: Fixes' },
  { type: 'revert', hidden: true, section: ':rewind: Revert' },
  { type: 'test', hidden: true, section: ':robot: Testing' },
  { type: 'docs', hidden: true, section: ':book: Documentation' },
  { type: 'build', hidden: true, section: ':hammer: Build' },
  {
    type: 'refactor',
    hidden: true,
    section: ':recycle: Code Improvements'
  },
  { type: 'perf', hidden: true, section: ':zap: Performance Updates' },
  { type: 'style', section: ':nail_care: Formatting & Structure' },
  {
    type: 'ci',
    section: ':truck: Continuous Integration & Deployment'
  },
  { type: 'chore', section: ':pencil2: Housekeeping' },
  { type: 'wip', hidden: true }
]
```

thus inadvertently hiding commit types from the `CHANGELOG`.

**To fix this issue, `grease` should override `types` after merging configs and assume that any type without a `section` property should be hidden. This means the `hidden` field only needs to be specified for clarity**:

```typescript
  args = merge({}, defaults, args)
  args.types = args.types?.map(type => ({ ...type, hidden: !type.section }))
```

With this fix, the `CHANGELOG` includes all entries for the specified version:

```zsh
▲ (FLEXBOOK) FLDV-P010 [next] yarn release:grease --release-as=patch --dry-run
! starting release workflow @flex-development/grease [dry=true]
✔ Running lifecycle script "prerelease"
ℹ - execute command: "yarn workspace @flex-development/grease test --no-cache"
✔ bumping version in package.json from 1.1.0 to 1.1.1
✔ outputting changes to CHANGELOG.md

---
### [1.1.1](https://github.com/flex-development/grease/compare/grease@1.1.0...grease@1.1.1) (2021-09-18)


### :pencil2: Housekeeping

* add eslint base config ([8de32f3](https://github.com/flex-development/grease/commit/8de32f3e4dd32c7f1cea85c9943b7ec88a1d7153))
* refine commit scopes ([ae15022](https://github.com/flex-development/grease/commit/ae15022381725668a773ab09c5c38b784e7cd83d))
* run `yarn set version berry` ([f94a4e1](https://github.com/flex-development/grease/commit/f94a4e162dbc3de7038b1bdb79573e25427baa5d))
* **typescript:** add typings for `shell-quote@1.7.2` ([70156cb](https://github.com/flex-development/grease/commit/70156cb7751374830f5c74d5b3c9d9f58c4b3e6a))
* update environment variables ([3f55036](https://github.com/flex-development/grease/commit/3f55036bf343aa8a3079294d3ebb4158cb4068e4))
* use yarn workspace based commit scopes ([889322b](https://github.com/flex-development/grease/commit/889322b2d5224a58dbde522f81cc3177da981c86))


### :recycle: Code Improvements

* **grease:** `CreateReleaseDTO#toString` ([0347d26](https://github.com/flex-development/grease/commit/0347d262a903f14692469da75974557dfced1f38))
* **grease:** integrate `@flex-development/log@2.0.0` ([787ac1d](https://github.com/flex-development/grease/commit/787ac1db313ca8ee3e0cab961aee2ab91b19531b))


### :robot: Testing

* **grease:** update changelog fixtures ([aabe44f](https://github.com/flex-development/grease/commit/aabe44fca6f2bf24060724cda5548793b58b080e))
* **grease:** use `DTO` fixture in `IsPath` functional test ([5150d51](https://github.com/flex-development/grease/commit/5150d511e12fed11779c647ceb6e5d671f322e8e))


### :hammer: Build

* **deps-dev:** drop `@types/yargs-unparser@2.0.1` ([e6adbfa](https://github.com/flex-development/grease/commit/e6adbfa7692dfdf0c4eae0b5085bbf6699e99c82))
* **deps-dev:** move `@types/lodash@4.14.172` to `devDependencies` ([4cebed3](https://github.com/flex-development/grease/commit/4cebed3e1c724fc848761b7e4a3226facf1d8737))
* **deps-dev:** move `@types/shelljs@0.8.9` to `devDependencies` ([283bc29](https://github.com/flex-development/grease/commit/283bc29413db865a35026d1c446c091525571e3b))
* **deps-dev:** use `@types/faker@5.5.8` ([ca4fb8e](https://github.com/flex-development/grease/commit/ca4fb8e7de0adfdaa3f0d3843898f1dcff7e3c90))
* **deps-dev:** use `@types/node@16.9.2` ([c0c1cd1](https://github.com/flex-development/grease/commit/c0c1cd1cabb25e8e21c662fcb1d839eb29eb582a))
* **deps-peer:** add `@types/node>=15` ([5ca3485](https://github.com/flex-development/grease/commit/5ca3485335012d0eb6660619a3e51456068bb8c0))
* **deps-peer:** add `@types/shelljs@0.8.9` ([60e82ea](https://github.com/flex-development/grease/commit/60e82ea33ea184c4751289cbdc449b0300e7f0fd))
* **deps:** add `@flex-development/log@2.0.0` ([1bc20c4](https://github.com/flex-development/grease/commit/1bc20c4f8701510ec604ece4ca5dce211296ec45))
* **deps:** add `shell-quote@1.7.2` ([6e72cd6](https://github.com/flex-development/grease/commit/6e72cd6782d057ef797d9b5d555edb702ed8828c))
* **deps:** use `isomorphic-git@1.10.0` ([3284bf4](https://github.com/flex-development/grease/commit/3284bf43ac5dd660374a629d1a872d56f74943dc))
* **deps:** use `lodash.*` deps ([4a9dc84](https://github.com/flex-development/grease/commit/4a9dc841d2e78259a1646da37904e91d368fcafb))
* run local integrity check ([0ab9086](https://github.com/flex-development/grease/commit/0ab9086048c10b77f041eb85819ee2a4c1cfddab))
* update build output ([3d97289](https://github.com/flex-development/grease/commit/3d972890cfa69e9f9622a54dc5371930d7079131))
---
``` 

[1]: https://lodash.com/docs/4.17.15#merge
